File: python/218.py
import heapq
import collections


class Solution(object):
    def getSkyline(self, buildings):
        """
        :type buildings: List[List[int]]
        :rtype: List[List[int]]
        """
        l = []
        for xi, xj, y in buildings:
            l.append((xi, -y))  # higher enter first
            l.append((xj, y))  # lower exit first
        l.sort()
        h = []  # max heap for height
        rm = collections.Counter()
        res = []
        for x, y in l:
            if y < 0:  # enter
                heapq.heappush(h, y)  # min 2 max
            else:  # mark as removed
                rm[y] += 1
            while h and rm[-h[0]]:
                rm[-h[0]] -= 1
                heapq.heappop(h)
            mxh = -h[0] if h else 0
            if not res or res[-1][1] != mxh:
                res.append((x, mxh))
        return res

sol = Solution()
# print(sol.getSkyline([[1, 2, 1], [1, 2, 2], [1, 2, 3]]))
print(sol.getSkyline([[0, 1, 3]]))

File: cpp/218.cpp
#include <algorithm>
#include <vector>
#include <set>
#include <utility>
using namespace std;

class Solution {
public:
  vector<pair<int, int>> getSkyline(vector<vector<int>> &buildings) {
    vector<pair<int, int>> l;
    for (auto &b : buildings) {
      l.push_back(make_tuple(b[0], -b[2]));
      l.push_back(make_tuple(b[1], b[2]));
    }
    sort(l.begin(), l.end());
    multiset<int> h{0};
    vector<pair<int, int>> res;
    for (auto &t : l) {
      if(t.second < 0)
        h.insert(-t.second);
      else
        h.erase(h.find(t.second));
      int mxh = *(h.rbegin());
        if (res.size() == 0 || res.back().second != mxh)
          res.push_back(make_pair(t.first, mxh));
      }
      return res;
  }
};
